Index: coregrind/m_errormgr.c
===================================================================
--- coregrind/m_errormgr.c	(revision 10974)
+++ coregrind/m_errormgr.c	(working copy)
@@ -169,6 +169,11 @@
    return n_errs_found;
 }
 
+void VG_(set_n_errs_found)( UInt n_errs )
+{
+   n_errs_found = n_errs;
+}
+
 /*------------------------------------------------------------*/
 /*--- Suppression type                                     ---*/
 /*------------------------------------------------------------*/
