Index: include/vki/vki-x86-linux.h
===================================================================
--- include/vki/vki-x86-linux.h	(revision 11055)
+++ include/vki/vki-x86-linux.h	(working copy)
@@ -481,11 +481,11 @@
 
 /* provoke compile error for invalid uses of size argument */
 #define _VKI_IOC_TYPECHECK(t) \
-	((sizeof(t) == sizeof(t[1]) && \
-	  sizeof(t) < (1 << _VKI_IOC_SIZEBITS)) \
-	 ? sizeof(t) \
-	 : /*cause gcc to complain about division by zero*/(1/0) )
+    (sizeof(t) / \
+     (sizeof(t) == sizeof(t[1]) && \
+      sizeof(t) < (1 << _VKI_IOC_SIZEBITS)))
 
+
 /* used to create numbers */
 #define _VKI_IO(type,nr)	_VKI_IOC(_VKI_IOC_NONE,(type),(nr),0)
 #define _VKI_IOR(type,nr,size)	_VKI_IOC(_VKI_IOC_READ,(type),(nr),(_VKI_IOC_TYPECHECK(size)))
