Index: include/pub_tool_libcassert.h
===================================================================
--- include/pub_tool_libcassert.h	(revision 10880)
+++ include/pub_tool_libcassert.h	(working copy)
@@ -31,18 +31,20 @@
 #ifndef __PUB_TOOL_LIBCBASSERT_H
 #define __PUB_TOOL_LIBCBASSERT_H
 
-#define tl_assert(expr)                                                 \
-  ((void) ((expr) ? 0 :                                                 \
-           (VG_(assert_fail) (/*isCore?*/False, #expr,                  \
-                              __FILE__, __LINE__, __PRETTY_FUNCTION__,  \
-                              ""),                                      \
+#define tl_assert(expr)                                               \
+((void) ((expr) ? 0 :                                                 \
+         (VG_(assert_fail) (/*isCore?*/False, (Char*)#expr,           \
+                              (Char*)__FILE__, __LINE__,              \
+                              (Char*)__PRETTY_FUNCTION__,             \
+                              ""),                                    \
                               0)))
 
-#define tl_assert2(expr, format, args...)                               \
-  ((void) ((expr) ? 0 :                                                 \
-           (VG_(assert_fail) (/*isCore?*/False, #expr,                  \
-                              __FILE__, __LINE__, __PRETTY_FUNCTION__,  \
-                              format, ##args),                          \
+#define tl_assert2(expr, format, args...)                             \
+  ((void) ((expr) ? 0 :                                               \
+           (VG_(assert_fail) (/*isCore?*/False, (Char*)#expr,         \
+                              (Char*)__FILE__, __LINE__,              \
+                              (Char*)__PRETTY_FUNCTION__,             \
+                              (const HChar*)format, ##args),          \
                               0)))
 
 __attribute__ ((__noreturn__))
Index: include/pub_tool_errormgr.h
===================================================================
--- include/pub_tool_errormgr.h	(revision 10880)
+++ include/pub_tool_errormgr.h	(working copy)
@@ -62,6 +62,14 @@
 Char*       VG_(get_error_string)  ( Error* err );
 void*       VG_(get_error_extra)   ( Error* err );
 
+
+typedef enum {
+  ERROR_IS_RECORDED,
+  ERROR_IS_SUPPRESSED,
+  ERROR_IS_IGNORED_AS_DUPLICATE,
+  ERROR_IS_IGNORED_DUE_TO_LIMIT
+} RecordErrorReturnType;
+
 /* Call this when an error occurs.  It will be recorded if it hasn't been
    seen before.  If it has, the existing error record will have its count
    incremented.
@@ -72,8 +80,10 @@
 
    If no 'a', 's' or 'extra' of interest needs to be recorded, just use
    NULL for them.  */
-extern void VG_(maybe_record_error) ( ThreadId tid, ErrorKind ekind,
-                                      Addr a, Char* s, void* extra );
+extern RecordErrorReturnType VG_(maybe_record_error) ( ThreadId tid, 
+                                                       ErrorKind ekind,
+                                                       Addr a, Char* s, 
+                                                       void* extra );
 
 /* Similar to VG_(maybe_record_error)(), except this one doesn't record the
    error -- useful for errors that can only happen once.  The errors can be
Index: include/vki/vki-x86-linux.h
===================================================================
--- include/vki/vki-x86-linux.h	(revision 10880)
+++ include/vki/vki-x86-linux.h	(working copy)
@@ -479,12 +479,15 @@
 	 ((nr)   << _VKI_IOC_NRSHIFT) | \
 	 ((size) << _VKI_IOC_SIZESHIFT))
 
-/* provoke compile error for invalid uses of size argument */
-extern unsigned int __vki_invalid_size_argument_for_IOC;
+/* The trunk version of the code below intentionally leads to compile-time
+ * error if anything is wrong with type "t".
+ * However, the original code is not C++-compiler-friendly.
+ *
+ * As a workaround, we produce a "division by zero" warning if anything is wrong
+ * with 't'. */
 #define _VKI_IOC_TYPECHECK(t) \
-	((sizeof(t) == sizeof(t[1]) && \
-	  sizeof(t) < (1 << _VKI_IOC_SIZEBITS)) ? \
-	  sizeof(t) : __vki_invalid_size_argument_for_IOC)
+	(sizeof(t) / (sizeof(t) == sizeof(t[1]) && \
+	  sizeof(t) < (1 << _VKI_IOC_SIZEBITS)))
 
 /* used to create numbers */
 #define _VKI_IO(type,nr)	_VKI_IOC(_VKI_IOC_NONE,(type),(nr),0)
Index: include/pub_tool_basics.h
===================================================================
--- include/pub_tool_basics.h	(revision 10880)
+++ include/pub_tool_basics.h	(working copy)
@@ -192,16 +192,19 @@
    }
    SysRes;
 #elif defined(VGO_darwin)
+  typedef
+  enum {
+     SysRes_MACH=40,  // MACH, result is _wLO
+     SysRes_MDEP,     // MDEP, result is _wLO
+     SysRes_UNIX_OK,  // UNIX, success, result is _wHI:_wLO
+     SysRes_UNIX_ERR  // UNIX, error,   error  is _wHI:_wLO
+  } ret_mode_t;
+
 typedef
    struct {
       UWord _wLO;
       UWord _wHI;
-      enum { 
-         SysRes_MACH=40,  // MACH, result is _wLO
-         SysRes_MDEP,     // MDEP, result is _wLO
-         SysRes_UNIX_OK,  // UNIX, success, result is _wHI:_wLO
-         SysRes_UNIX_ERR  // UNIX, error,   error  is _wHI:_wLO
-      } _mode;
+      ret_mode_t _mode;
    }
    SysRes;
 #else
Index: coregrind/m_errormgr.c
===================================================================
--- coregrind/m_errormgr.c	(revision 10880)
+++ coregrind/m_errormgr.c	(working copy)
@@ -626,7 +626,7 @@
 /* Top-level entry point to the error management subsystem.
    All detected errors are notified here; this routine decides if/when the
    user should see the error. */
-void VG_(maybe_record_error) ( ThreadId tid, 
+RecordErrorReturnType VG_(maybe_record_error) ( ThreadId tid, 
                                ErrorKind ekind, Addr a, Char* s, void* extra )
 {
           Error  err;
@@ -673,7 +673,7 @@
          VG_(umsg)("\n");
          stopping_message = True;
       }
-      return;
+      return ERROR_IS_IGNORED_DUE_TO_LIMIT;
    }
 
    /* After M_COLLECT_ERRORS_SLOWLY_AFTER different errors have
@@ -720,8 +720,7 @@
             p->next      = errors;
             errors       = p;
 	 }
-
-         return;
+         return p->supp ? ERROR_IS_SUPPRESSED : ERROR_IS_IGNORED_AS_DUPLICATE;
       }
       p_prev = p;
       p      = p->next;
@@ -778,10 +777,12 @@
       pp_Error( p, /*allow_db_attach*/True );
       /* update stats */
       n_errs_shown++;
+      return ERROR_IS_RECORDED;
    } else {
       n_supp_contexts++;
       n_errs_suppressed++;
       p->supp->count++;
+      return ERROR_IS_SUPPRESSED;
    }
 }
 
